src/contracts/core/StakeRootCompendium.sol

```diff
@@ -0,0 +1,618 @@
+            strategiesAndMultipliers.set(
+                address(_strategiesAndMultipliers[i].strategy), uint256(_strategiesAndMultipliers[i].multiplier)
+            );
+        }
+        uint256 numStrategiesAfter = strategiesAndMultipliers.length();
+
+        DepositInfo storage depositInfo = depositInfos[operatorSet.avs][operatorSet.operatorSetId];
+
+        // make sure they have enough to pay for MIN_PREPAID_PROOFS
+        require(depositInfo.balance >= minDepositBalance(numStrategiesAfter), InsufficientDepositBalance());
+        // note that they've shown increased demand for proofs
+        depositInfo.lastDemandIncreaseTimestamp = uint32(block.timestamp);
+
+        // only adding new strategies to count
+        _updateTotalStrategies(numStrategiesBefore, numStrategiesAfter);
+    }
+
+    /// @inheritdoc IStakeRootCompendium
+    function removeStrategiesAndMultipliers(uint32 operatorSetId, IStrategy[] calldata strategies) external {
+        OperatorSet memory operatorSet = OperatorSet({avs: msg.sender, operatorSetId: operatorSetId});
+        // update the deposit balance for the operator set whenever number of strategies is changed
+        _updateDepositInfo(operatorSet);
+
+        EnumerableMap.AddressToUintMap storage strategiesAndMultipliers =
+            operatorSetToStrategyAndMultipliers[operatorSet.avs][operatorSetId];
+
+        // note below either all strategies are removed or none are removed and transaction reverts
+        uint256 numStrategiesBefore = strategiesAndMultipliers.length();
+        for (uint256 i = 0; i < strategies.length; i++) {
+            require(strategiesAndMultipliers.remove(address(strategies[i])), NonexistentStrategy());
+        }
+
+        _updateTotalStrategies(numStrategiesBefore, numStrategiesBefore - strategies.length);
+    }
+
+    /// @inheritdoc IStakeRootCompendium
+    function setOperatorSetExtraData(uint32 operatorSetId, bytes32 extraData) external {
+        require(_isInStakeTree(OperatorSet({avs: msg.sender, operatorSetId: operatorSetId})), NotInStakeTree());
+        operatorSetExtraDatas[msg.sender][operatorSetId] = extraData;
+    }
+
+    /// @inheritdoc IStakeRootCompendium
+    function setOperatorExtraData(uint32 operatorSetId, address operator, bytes32 extraData) external {
+        require(_isInStakeTree(OperatorSet({avs: msg.sender, operatorSetId: operatorSetId})), NotInStakeTree());
+        operatorExtraDatas[msg.sender][operatorSetId][operator] = extraData;
+    }
+
+    /// @inheritdoc IStakeRootCompendium
+    function withdraw(uint32 operatorSetId, uint256 amount) external payable returns (uint256) {
+        OperatorSet memory operatorSet = OperatorSet({avs: msg.sender, operatorSetId: operatorSetId});
+        require(canWithdrawDepositBalance(operatorSet), OperatorSetNotOldEnough());
+
+        // debt any pending charge
+        uint256 balance = _updateDepositInfo(operatorSet);
+        if (balance < MIN_BALANCE_THRESHOLD + amount) {
```

Comment:
IMO we should have this function specifically prevent you from withdrawing `amount` if you go less than `MIN_BALANCE_THRESHOLD` - high potential for footguns by AVSs. Would be better to have a separate, voluntary, `removeFromStakeTreeAndWithdraw` func

Reply:
great point. agree